custom/aaa/tasks.py

```diff
@@ -0,0 +1,59 @@
+from __future__ import absolute_import, unicode_literals
+
+from datetime import datetime
+
+from celery.task import task
+from django.db import connections
+
+from custom.aaa.models import AggregationInformation, Child
+
+
+def update_table(domain, slug, method):
+    window_start = AggregationInformation.objects.filter(
+        step=slug, aggregation_window_end__isnull=False
+    ).order_by('-created_at').values_list('aggregation_window_end').first()
+    if window_start is None:
+        window_start = datetime(1900, 1, 1)
+    else:
+        window_start = window_start[0]
+
+    window_end = datetime.utcnow()
```

Comment:
Looks like this is using date-based aggregation windows rather than sequence-id based with the intricate locking procedure described in [the citus article](https://www.citusdata.com/blog/2018/06/14/scalable-incremental-data-aggregation/). Did you decide against the approach outlined in that article, or are you using a variant that does not have all the same characteristics?

Reply:
Yeah. I'm using dates because we insert the current time in inserted_at and we do not have a sequential ID. inserted_at is the only increasing ID we have for now. (the primary key is doc_id which is random uuid)

Reply:
From the article:

> There might still be in-progress writes to the events table that were assigned lower sequence values, but are not yet visible when the aggregation runs

Since this is not using the locking trick, I think the same applies to dates the way they're used here, which means some rows may be missed by the aggregation. Also I think small date differences between servers could cause issues.

Reply:
Agreed about the small date differences between different dbs making this messy. Its possible that we could define some amount of drift that is expected and have the aggregation window end there. So the end date is always `now() - 1 minute`. That would solve this for machines with drift of up to one minute. Currently this is not an issue because it will all be on one server. It will be good to look into what sort of guarantees Citus can provide about synchronization of nodes (thought currently that utcnow does come from the celery task in python. Possible that we could use `now()` ad defined by postgres).

I am planning to use the locking trick as commented in line 28. From my understanding the workflow would be

1. Set window's end time
2. Lock the table for writes
3. Release lock.
    * all writes to the database that were inserted_at < time at point 1 are guaranteed to have been written.
4. Agg data that has been inserted up until point 1.

Reply:
> I am planning to use the locking trick as commented in line 28.

Ah, I didn't realize that comment was a TODO item.

Using dates rather than a sequence still seems to me more complex and harder to reason about. `datetime.utcnow()` is not monotonically increasing (clock drift is real, and the NTP service adjusts system time routinely). See [SO answer](https://stackoverflow.com/a/3527632/10840) for some examples of how complex a subject this is. If I were doing this I would use a sequence as described in the article.

Reply:
Yeah. That does make a lot of sense. We don't currently have a way to add sequences to a UCR data source, but I imagine it would not be difficult. I would need to think more about adding it to case UCRs since those are updated and not just inserted (like the 99% case for forms).

I have [removed the where clauses](https://github.com/dimagi/commcare-hq/pull/23243/commits/f702bd968604037b5b24655b89d85f17400f4a1b) for now while I think about this. This means that the aggregation_window_start and window_end are not used and each aggregation is a full aggregation. I have kept those database as they will still be useful concepts for the monthly location rollup tables that are coming next.